ui/raidboss/data/06-ew/raid/p5s.ts

```diff
@@ -60,6 +71,56 @@ const triggerSet: TriggerSet<Data> = {
       netRegex: NetRegexes.startsUsing({ id: '76F3', source: 'Proto-Carbuncle', capture: false }),
       response: Responses.aoe(),
     },
+    {
+      id: 'P5S Ruby 1 Topaz Stone Collect',
+      type: 'Ability',
+      netRegex: NetRegexes.ability({ id: '76FE', source: 'Proto-Carbuncle' }),
+      condition: (data) => !data.isRuby1Done,
+      run: (data, matches) => {
+        data.ruby1TopazStones.push(matches);
+      },
+    },
+    {
+      id: 'P5S Ruby 1 Topaz Stones',
+      type: 'Ability',
+      netRegex: NetRegexes.ability({ id: '76FD', source: 'Proto-Carbuncle', capture: false }),
+      condition: (data) => !data.isRuby1Done,
+      delaySeconds: 0.3, // allow collector to finish
+      infoText: (data, _matches, output) => {
+        const safeQuadrants = Object.assign({}, directions);
+        for (const stone of data.ruby1TopazStones)
+          delete safeQuadrants[convertCoordinatesToDirection(parseFloat(stone.targetX), parseFloat(stone.targetY))];
+
+        const safe: string[] = Object.keys(safeQuadrants);
```

Comment:
It's a little unfortunate to push everything through strings here when they could have a better type.  You could do something like this, so that `safe` is more strict than just a `string[]`.

```suggestion
        const safeQuadrants = new Set(directions);
        for (const stone of data.ruby1TopazStones)
          safeQuadrants.delete(convertCoordinatesToDirection(parseFloat(stone.targetX), parseFloat(stone.targetY)));

        const safe = Array.from(safeQuadrants);
```

if you also did this, so you'd have an array to make a set from:

```diff
diff --git a/ui/raidboss/data/06-ew/raid/p5s.ts b/ui/raidboss/data/06-ew/raid/p5
s.ts
index 3a0f3906f..2c87ff8fd 100644
--- a/ui/raidboss/data/06-ew/raid/p5s.ts
+++ b/ui/raidboss/data/06-ew/raid/p5s.ts
@@ -9,23 +9,19 @@ import { TriggerSet } from '../../../../../types/trigger';

 // TODO: Callout safe quadrant/half for Venom Pool with Crystals

-const directions = {
-  'NE': true,
-  'SE': true,
-  'SW': true,
-  'NW': true,
-};
+export const directions = ['NW', 'NE', 'SE', 'SW'] as const;
+export type IntercardDir = typeof directions[number];

 export interface Data extends RaidbossData {
   target?: string;
   topazClusterCombatantIdToAbilityId: { [id: number]: string };
-  topazRays: { [time: number]: (keyof typeof directions)[] };
+  topazRays: { [time: number]: IntercardDir[] };
   clawCount: number;
   ruby1TopazStones: NetMatches['Ability'][];
   isRuby1Done: boolean;
 }

-export const convertCoordinatesToDirection = (x: number, y: number): keyof typeof directions => {
+export const convertCoordinatesToDirection = (x: number, y: number): IntercardDir => {
   if (x > 100)
     return y < 100 ? 'NE' : 'SE';
   return y < 100 ? 'NW' : 'SW';

@@ -206,9 +202,9 @@ const triggerSet: TriggerSet<Data> = {
         }
       },
       infoText: (data, _matches, output) => {
-        const remainingDirections: { [index: string]: Set<keyof typeof directions> } = {};
+        const remainingDirections: { [index: string]: Set<IntercardDir> } = {};
         for (const [index, directions] of Object.entries(data.topazRays)) {
-          remainingDirections[index] = new Set(['NE', 'SE', 'SW', 'NW']);
+          remainingDirections[index] = new Set(directions);
           for (const direction of directions)
             remainingDirections[index]?.delete(direction);
         }
```

Reply:
> It's a little unfortunate to push everything through strings here when they could have a better type.

Agree - I think this is much cleaner.